GraphComponent : Use DirtyPropagationScope for parent changes

danieldresser-ie opened the pull request:
This is a core thing that I don't know all the details of, so you definite may want to change up some of the details.  But I have got a nice simple test that demonstrates that misuse of undo scopes can currently trigger an exception that results in a mismatched DirtyPropagationScope.

I also observed that you have now added a DirtyPropagationScope to `~GraphComponent`.  Since you suggested the original reason for not doing this around parentChanging/parentChanged in GraphComponent was so that GraphComponent wouldn't know about dirty propagation, the fact that it now does this in one place strongly suggests that GraphComponent should always take care of it.

My naive fix does seem to address the issue.  Some questions about it, that you would probably know more than I would about:
* should we release the DirtyPropagationScope before `MemberSignals::emitLazily( ..., parentChangedSignal )`?  Not sure what makes more sense, but in this naive version, I've implicitly changed the order from what it was before
* should we avoid creating a DirtyPropagationScope when emitParentChanged == False?  Again, not sure what makes sense, but that's the way it was before

danieldresser-ie commented:
One additional question : why did CropWindowTool start failing?  The fix seems kind of reasonable, but I'm not sure why this changed.

johnhaddon commented:
Thanks Daniel! Good job hunting this down!

> should we release the DirtyPropagationScope before MemberSignals::emitLazily( ..., parentChangedSignal )? Not sure what makes more sense, but in this naive version, I've implicitly changed the order from what it was before

I think the way you have it makes the most sense. We don't document whether or not `childAdded` and `childRemoved` slots are allowed to modify the graph in the same way as `plugInputChanged` may, but I suspect that they _do_ in practice, and it seems reasonable that they might want to. Postponing `plugDirtiedSignal` emission till after any such rewiring is preferable, so that any observers that trigger computes will do so only when the graph has reached its final state. There's some discussion of this related to `plugInputChanged` [here](https://github.com/GafferHQ/gaffer/blob/main/include/Gaffer/Node.h#L87-L118).

> should we avoid creating a DirtyPropagationScope when emitParentChanged == False? 

It seems reasonable to create the DirtyPropagationScope even in this case. We're still emitting `childRemovedSignal()`, which could potentially lead to rewiring, and again we'd want to delay dirty signalling until that was done.

> why did CropWindowTool start failing?

I think this is because `Node::parentChanging()` (Tools are Nodes) removes all connections to the node. Previously, that would have emitted `dirtiedSignal()` for those plugs immediately, while the node still had a parent. Now that emission is deferred until after unparenting is complete, at which point the tool can't find it's ancestor View. I am a bit concerned about this - I already [mitigated against the same problem](https://github.com/GafferHQ/gaffer/blob/main/src/GafferUI/View.cpp#L94-L98) before, because I think it's better if Tools aren't exposed to this state. I don't have any better ideas right now though, so it's either run with the workaround you have, or spend a bit more time scratching heads...

danieldresser-ie commented:
I've now added a revert for the CropWindowTool fix, and added your proposed alternate fix ... seems OK so far, though I really couldn't claim to know what it might interact with.

johnhaddon commented:
Since I agreed to pick this one up, I've pushed a new version to this PR which does the following :

- Removes the reverted CropWindowTool fix.
- Tweaks the naming of the new Trackable method, binds it to Python and adds a test.
- Updates Changes.md.
- Adds an unrelated comment to Signals.h to document some of my recent thinking.

@danieldresser-ie, if you could give this a quick once over and merge it if it looks OK, that would be grand.
